contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,335 @@
+            0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8, // USDC.e
+            0xFd086bC7CD5C481DCC9C85ebE478A1C0b69FCbb9 // USDT
+        ];
+    }
+
+    function decodePublishPayload(
+        bytes32 payload
+    ) public pure returns (EncodedIntent memory) {
+        return
+            EncodedIntent({
+                sourceChainIndex: uint8(payload[0]), // uint8
+                destinationChainIndex: uint8(payload[1]), // uint8
+                rewardTokenIndex: uint8(payload[2]), // uint8
+                // Reads bytes 3 to 8 and converts them to uint48
+                rewardAmount: uint48(_extractUint(payload, 3, 8)), // uint48
+                routeTokenIndex: uint8(payload[9]), // uint8
+                // Reads bytes 10 to 15 and converts them to uint48
+                routeAmount: uint48(_extractUint(payload, 10, 15)), // uint48
+                expiryDuration: uint24(_extractUint(payload, 16, 18)) // uint24
+            });
+    }
+
+    function decodeFulfillPayload(
+        bytes32 payload
+    ) public pure returns (EncodedFulfillment memory) {
+        return
+            EncodedFulfillment({
+                sourceChainIndex: uint8(payload[0]), // uint8
+                destinationChainIndex: uint8(payload[1]), // uint8
+                routeTokenIndex: uint8(payload[2]), // uint8
+                // Reads bytes 10 to 15 and converts them to uint48
+                routeAmount: uint48(_extractUint(payload, 3, 8)), // uint48
+                proveType: uint8(payload[9]),
+                claimant: address(uint160(_extractUint(payload, 10, 29))) // uint48
+            });
+    }
+
+    // ======================== Internal Functions ========================
+
+    function _fundIntent(Intent memory intent) internal {
+        address funder = msg.sender;
+        TokenAmount[] memory tokens = intent.reward.tokens;
+
+        // Get vault address from intent
+        address vault = INTENT_SOURCE.intentVaultAddress(intent);
+
+        // Cache tokens length
+        uint256 rewardsLength = tokens.length;
+
+        // Iterate through each token in the reward structure
+        for (uint256 i; i < rewardsLength; ++i) {
+            // Get token address and required amount for current reward
+            address token = tokens[i].token;
+            uint256 amount = tokens[i].amount;
+            uint256 balance = IERC20(token).balanceOf(vault);
+
+            // Only proceed if vault needs more tokens and we have permission to transfer them
+            if (amount > balance) {
+                // Calculate how many more tokens the vault needs to be fully funded
+                uint256 remainingAmount = amount - balance;
+
+                // Check how many tokens this contract is allowed to transfer from funding source
+                uint256 allowance = IERC20(token).allowance(
+                    funder,
+                    address(this)
+                );
+
```

Comment:
Please add the funder balance check as in the latest version of IntentSource/Vault. I think we should make it a library lol